### PR TITLE
feat: NPU (Ascend 910B3) support for RFdiffusion inference

### PR DESCRIPTION
**Pull request by xuejiakn**
## Description

This PR adapts RFdiffusion to run on **Huawei Ascend NPU (Ascend 910B3)** with `torch_npu`, enabling protein structure generation on NPU hardware.

### Background

RFdiffusion depends on PyTorch + DGL + e3nn + SE3Transformer, all originally CUDA-only. This PR replaces CUDA-specific calls with NPU-compatible alternatives and patches operators not supported on NPU.

### Changes (8 files, +114 -16 lines)

| File | Change |
|------|--------|
| `rfdiffusion/__init__.py` | `allow_internal_format=False` + `torch.cdist` NPU patch (manual `bmm + sqrt`) |
| `rfdiffusion/inference/model_runners.py` | Device: `torch.cuda` → `torch.npu` |
| `rfdiffusion/Track_module.py` | autocast: `torch.cuda.amp.autocast` → `torch.amp.autocast(device_type="npu")` |
| `scripts/run_inference.py` | NPU detection (NPU > CUDA > CPU), `torch.npu.empty_cache()`, device metadata |
| `SE3Transformer/model/basis.py` | nvtx three-tier fallback: CUDA → `record_function` → no-op |
| `SE3Transformer/model/layers/attention.py` | nvtx fallback + `e_dot_v` manual implementation |
| `SE3Transformer/model/layers/convolution.py` | nvtx fallback |
| `SE3Transformer/model/layers/norm.py` | nvtx fallback |

### Key Adaptations

1. **Device migration**: All `torch.cuda.*` calls replaced with `torch.npu.*` equivalents
2. **Internal format disabled**: `torch.npu.config.allow_internal_format = False` to fix `torch.cat` tensor format conflicts in SE3Transformer
3. **torch.cdist patch**: NPU does not support `torch.cdist` natively; patched with `||x1-x2|| = sqrt(||x1||² + ||x2||² - 2*x1·x2^T)`
4. **nvtx fallback**: `torch.cuda.nvtx.range` import succeeds but throws at runtime on NPU; three-tier fallback with `torch.autograd.profiler.record_function` for NPU profiling support
5. **DGL e_dot_v**: Ascend DGL SDDMM only supports `lhs_target=0`; replaced with manual `(key * query[dst]).sum(dim=-1)`

### Prerequisites

- DGL Ascend version with SpMM `copy_rhs` fix (see [dgl-ascend PR #24](https://gitcode.com/AI4Science/dgl-ascend/merge_requests/24))
- PyTorch 2.7.1+cpu + torch_npu 2.7.1.post4
- CANN 8.2.RC1, Ascend 910B3

### Verification

| Metric | NPU | CPU | Diff |
|--------|-----|-----|------|
| pLDDT (20-step deterministic) | 0.9885 | 0.9885 | 0.000005 |
| Inference time (20 steps) | 43.3s | 664.7s | 15.35x speedup |

- All 6 checkpoints verified (Base, ActiveSite, Complex, InpaintSeq, etc.)
- pLDDT > 0.95 for 50-residue designs (normal random mode)
- 20-step inference quality matches 200-step (consistent with paper)

## Type

- [x] New feature (NPU hardware support)
- [ ] Bug fix
- [ ] Performance optimization
- [ ] Documentation update
- [ ] Other

### CODE REVIEW

**Comment by rclune** [comment]
Thank you for this PR. 

There are newer version of RFdiffusion (most notably [RFD3](https://github.com/RosettaCommons/foundry/tree/production/models/rfd3)) whose dependencies are more up to date. Do you run into similar issues trying to RFD3 on the NPU resources you have access to?

File: scripts/run_inference.py [hunk 1]
```diff
@@ -18,6 +18,8 @@
 import re
 import os, time, pickle
 import torch
+import torch_npu
```
Comment (rclune) [inline scripts/run_inference.py:21]: Many users of RFD only have access to GPU/CPU resources. Adding this import statement not in a try/except will make it impossible to run on GPU systems. I suggest wrapping it like you do in `__init__.py`.

File: rfdiffusion/inference/model_runners.py [hunk 2]
```diff
@@ -48,8 +48,8 @@ def initialize(self, conf: DictConfig) -> None:
 
         """
         self._log = logging.getLogger(__name__)
-        if torch.cuda.is_available():
-            self.device = torch.device("cuda")
+        if torch.npu.is_available():
```
Comment (rclune) [inline rfdiffusion/inference/model_runners.py:51]: Please back in the check for `torch.cuda.is_available()` as a fallback. Right now only NPU is checked, so GPU users will be silently placed on CPUs. You can do something similar to what you have on lines 46-50 of `run_inference.py`.

File: rfdiffusion/Track_module.py [hunk 3]
```diff
@@ -233,7 +233,7 @@ def reset_parameter(self):
         nn.init.zeros_(self.embed_e1.bias)
         nn.init.zeros_(self.embed_e2.bias)
     
-    @torch.cuda.amp.autocast(enabled=False)
+    @torch.amp.autocast(device_type="npu", enabled=False)
```
Comment (rclune) [inline rfdiffusion/Track_module.py:236]: This hardcodes device_type="npu", so on CUDA systems this decorator no longer disables autocast, meaning torch.cuda.amp.autocast(enabled=False)'s effect is lost. Should be conditional on the active device.

File: env/SE3Transformer/se3_transformer/model/layers/attention.py [hunk 4]
```diff
@@ -78,7 +92,10 @@ def forward(
 
             with nvtx_range('attention dot product + softmax'):
                 # Compute attention weights (softmax of inner product between key and query)
-                edge_weights = dgl.ops.e_dot_v(graph, key, query).squeeze(-1)
```
Comment (rclune) [inline env/SE3Transformer/se3_transformer/model/layers/attention.py:81]: The removal of this for a manual implementation should be conditioned on the device type, similar to how the nvtx changes are.

**Review by rclune — CHANGES_REQUESTED** [review]
Thank you for this PR. I have some concerns around ensuring that the changes made will still allow run_inference to work with GPU/CUDA systems. Please make sure that any NPU-specific additions are wrapped in conditionals or try/except clauses to ensure that the GPU functionality is not changed.
